### PR TITLE
Support multiple chore processes consuming from the filesystem queue

### PR DESCRIPTION
**Pull request by obrie**
In order to give us more flexibility in how we configure and deploy chore as a filesystem sidecar queue, this adds support for being able to run multiple chore masters or multiple forked workers.  Currently this is supported for external services like SQS; however, we don't have graceful support for it in the filesystem.

Similar to https://github.com/Tapjoy/chore/pull/43, this gets us a lot more flexibility in deployment configuration without investing time into a Go-based filesystem queue sidecar.

## Design

Currently the filesystem queue already supports multiple consumers across different processes.  However, the problem we still need to solve is what happens to in-progress jobs when a forked worker or master process dies.  The current implementation assumes only 1 master is ever running at a time.  This is because, on startup, it will immediately move everything out of "in progress" and into "new".  This works in a single master scenario since you can assume no one is processing the stuff that's in progress.  However, in multi-master, some of those jobs may be expired / leftover from old killed processes and some may be actively getting processed by a live master.

To address this, we take a page out of SQS's book.  We'll only ever delete from "in progress" if the job was completed *or* the job has expired based on the queue's visibility timeout.  The former is the normal scenario.  In the latter, jobs will expire if either (a) they took too long to process or (b) the process that owned the job was killed.

To track expiration time, we put a timestamp into the job's filename for when it was moved into "in progress".  This is necessary since neither the creation, modification, or access times change when the file moves.  This expiration time gets removed when the file is moved back into "new".

In addition to this periodic cleanup process, additionally error handling is needed in order to account for scenarios where we are both completing a job and cleaning up the job simulatenously.

## Performance

From a performance perspective, this is negligible unless there is a large backlog of files in "new".  In that case, we only allow the "new" folder to be cleaned up at most once a second in order to avoid a performance impact.  Once nice thing from this is that it actually allows for some delay before the job gets retried (jobs only get retried once every `visibility timeout` seconds).

## Summary

In summary, we've made the filesystem queue behave much more in line with how SQS behaves -- allowing us to make more sense of how jobs will behave regardless of the queue selected.

## TODO

* [x] Verify locally

### CODE REVIEW

**Comment by obrie** [comment]
/to @jjrussell @dankleiman @gregorysabatino @geoffreyclark for early review

File: lib/chore/queues/filesystem/consumer.rb [hunk 1]
```diff
@@ -99,7 +106,9 @@ def consume
           Chore.logger.info "Starting consuming file system queue #{@queue_name} in #{self.class.queue_dir(queue_name)}"
           while running?
             begin
-              #TODO move expired job files to new directory?
+              # Move expired job files to new directory?
```
Comment (dankleiman) [inline lib/chore/queues/filesystem/consumer.rb]: nit: remove `?` in comment

**Comment by obrie** [comment]
I've updated the PR description.  This should be ready for full review.

**Comment by dankleiman** [comment]
LGTM!

File: lib/chore/consumer.rb [hunk 2]
```diff
@@ -21,11 +21,6 @@ def initialize(queue_name, opts={})
     def self.reset_connection!
     end
 
-    # Cleans up any resources that were left behind from prior instances of the
-    # chore process.  By default, this is a no-op.
-    def self.cleanup(queue)
-    end
-
```
Comment (gregorysabatino) [inline lib/chore/consumer.rb]: just curious, why are you removing this from the "base class"?
i get that we are now providing an implementation in filesystem consumer, but question maybe still relevant?

Comment (gregorysabatino) [reply lib/chore/consumer.rb]: oh, think i got it now, nevermind

File: lib/chore/queues/filesystem/consumer.rb [hunk 3]
```diff
@@ -99,7 +117,13 @@ def consume
           Chore.logger.info "Starting consuming file system queue #{@queue_name} in #{self.class.queue_dir(queue_name)}"
           while running?
             begin
-              #TODO move expired job files to new directory?
+              # Move expired job files to new directory (so long as enough time has
+              # passed since we last did this check)
+              if !@last_cleaned_at || (Time.now - @last_cleaned_at).to_i >= MINIMUM_EXPIRATION_CHECK_INTERVAL
```
Comment (gregorysabatino) [inline lib/chore/queues/filesystem/consumer.rb]: was it intended to have this `MINIMUM_` get overridden?  just an assumption i made based on naming of the constant.  not something i view as a blocker fer sher.

**Comment by gregorysabatino** [comment]
LGTM ; just 1 non-blocking question

File: lib/chore/queues/filesystem/consumer.rb (hunk 3 above)
Comment (obrie) [reply lib/chore/queues/filesystem/consumer.rb]: Good call, I'll rename this variable.  I could imagine some need for this to be configurable, but until that need is actually there, I would just leave this a constant.

Comment (obrie) [reply lib/chore/queues/filesystem/consumer.rb]: Actually, I think the name may be more accurate as-is (I'll spend a few minutes thinking aout it).  The code doesn't guarantee 1s intervals -- it just guarantees that there's at least 1s between checks.

Comment (obrie) [reply lib/chore/queues/filesystem/consumer.rb]: Thought about it more and decided to rename it, but leave in the comments indicating that this is more of a minimum value
